### ISSUE TITLE
API KEY Unique

### ISSUE THREAD

**Issue by dvdieu** [apache/apisix#11197]
### Description

API KEY can be use for purpose Authentication, to do that API KEY need require to unique  for all Consumer

**Comment by nitishfy** [comment]
The API_KEY is already unique for different consumers. No two users have the same api_key.

**Comment by dvdieu** [comment]
> The API_KEY is already unique for different consumers. No two users have the same api_key.

Hi,
I realy created 2 consumer with the same API KEY, you can try this

**Comment by kayx23** [comment]
The `username` should be unique

**Comment by dvdieu** [comment]
username required unique but API Key the same, because if API Key not unique for all user how about your detect user A with user B

**Comment by moonming** [comment]
> username required unique but API Key also, because if API Key not unique for all user how about your detect user A with user B.
> 
> What happens when the username is unique but the API Key is duplicate, then determining which user is calling is impossible

@dvdieu You are right, if there are duplicate API keys, it is impossible to identify different consumers.
I don't have a good proposal on how to solve it yet, because Apache APISIX uses etcd as storage instead of a relational database, and it's not easy to determine if there are duplicates. What do you think?

**Comment by dvdieu** [comment]
> You are right, if there are duplicate API keys, it is impossible to identify different consumers.
> I don't have a good proposal on how to solve it yet, because Apache APISIX uses etcd as storage instead of a relational database, and it's not easy to determine if there are duplicates. What do you think?

i think
 local client = etcd.client({host = etcd_host, port = etcd_port})
    local response, err = client:get(key)
    if err then
        return false
    end
    if response.count > 1 then
        return true
    else
        return false
    end
    
    for scaling we can using redis for check before insert to ectd

**Comment by moonming** [comment]
> > You are right, if there are duplicate API keys, it is impossible to identify different consumers.
> > I don't have a good proposal on how to solve it yet, because Apache APISIX uses etcd as storage instead of a relational database, and it's not easy to determine if there are duplicates. What do you think?
> 
> i think local client = etcd.client({host = etcd_host, port = etcd_port}) local response, err = client:get(key) if err then return false end if response.count > 1 then return true else return false end
> 
> ```
> for scaling we can using redis for check before insert to ectd
> ```

If this judgment logic can be added when adding API key instead of processing client requests, it is good for me

**Comment by dvdieu** [comment]
Please update this bug
